x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -0,0 +1,325 @@
+			SecretAccessKey: m.config.AwsSecretAccessKey,
+			SessionToken:    m.config.AwsSessionToken,
+		},
+	}
+	cfg.Region = m.config.AwsDefaultRegion
+
+	svcEC2 := ec2.New(cfg)
+	//Get a list of regions
+	regionsList, err := getRegions(svcEC2)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getRegions failed"))
+		return
+	}
+
+	for _, regionName := range regionsList {
+		cfg.Region = regionName
+		svcEC2 := ec2.New(cfg)
+		instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getInstancesPerRegion failed, skipping region "+regionName))
+			continue
+		}
+
+		svcCloudwatch := cloudwatch.New(cfg)
+		for _, instanceID := range instanceIDs {
+			init := true
+			getMetricDataOutput := &cloudwatch.GetMetricDataOutput{NextToken: nil}
+			for init || getMetricDataOutput.NextToken != nil {
+				init = false
+				output, err := getMetricDataPerRegion(instanceID, getMetricDataOutput.NextToken, svcCloudwatch)
+				if err != nil {
+					report.Error(errors.Wrap(err, "getMetricDataPerRegion failed"))
+					return
+				}
+				getMetricDataOutput.MetricDataResults = append(getMetricDataOutput.MetricDataResults, output.MetricDataResults...)
+			}
+			reportCloudWatchEvents(getMetricDataOutput, instanceID, instancesOutputs[instanceID], regionName, report)
+		}
+	}
+}
+
+// MockFetch methods implements the data gathering and data conversion using MockEC2Client and MockCloudWatchClient
+func (m *MetricSet) MockFetch(report mb.ReporterV2) {
+	svcEC2Mock := &MockEC2Client{}
+	instanceIDs, instancesOutputs, err := getInstancesPerRegion(svcEC2Mock)
+	if err != nil {
+		report.Error(errors.Wrap(err, "getInstancesPerRegion failed"))
+	}
+
+	svcCloudwatchMock := &MockCloudWatchClient{}
+	for _, instanceID := range instanceIDs {
+		getMetricDataOutput, err := getMetricDataPerRegion(instanceID, nil, svcCloudwatchMock)
+		if err != nil {
+			report.Error(errors.Wrap(err, "getMetricDataPerRegion failed"))
+		}
+		reportCloudWatchEvents(getMetricDataOutput, instanceID, instancesOutputs[instanceID], "us-west-1", report)
+	}
+}
+
+func getRegions(svc ec2iface.EC2API) (regionsList []string, err error) {
+	input := &ec2.DescribeRegionsInput{}
+	req := svc.DescribeRegionsRequest(input)
+	output, err := req.Send()
+	if err != nil {
+		logp.Error(errors.Wrap(err, "Failed DescribeRegions"))
+		return
+	}
+	for _, region := range output.Regions {
+		regionsList = append(regionsList, *region.RegionName)
+	}
+	return
+}
+
+func reportCloudWatchEvents(getMetricDataOutput *cloudwatch.GetMetricDataOutput, instanceID string, instanceOutput ec2.Instance, regionName string, report mb.ReporterV2) {
+	machineType, err := instanceOutput.InstanceType.MarshalValue()
+	if err != nil {
+		report.Error(errors.Wrap(err, "instance.InstanceType.MarshalValue failed"))
+	}
+
+	event := mb.Event{}
+	event.RootFields = common.MapStr{}
+	event.RootFields.Put("service.name", aws.ModuleName)
+	event.RootFields.Put("cloud.provider", "ec2")
```

Comment:
Looks like the above generated event is just out of sync?

Reply:
@ruflin sorry what do you mean by out of sync here? This function `reportCloudWatchEvents` is to report one event per one specific instanceID.

Reply:
All good, was related to my data.json comment. Ignore it :-)